README.md
DudelDu
=======
DudelDu is a simple audio/video streaming server using the SHOUTcast protocol.

<p>
<a href="https://void.devt.de/pub/dudeldu/coverage.txt"><img src="https://void.devt.de/pub/dudeldu/test_result.svg" alt="Code coverage"></a>
<a href="https://goreportcard.com/report/devt.de/krotik/dudeldu">
<img src="https://goreportcard.com/badge/devt.de/krotik/dudeldu?style=flat-square" alt="Go Report Card"></a>
<a href="https://godoc.org/devt.de/krotik/dudeldu">
<img src="https://godoc.org/devt.de/krotik/dudeldu?status.svg" alt="Go Doc"></a>
</p>

Features
--------
- Supports various streaming clients: <a href="http://www.videolan.org/vlc/download-windows.en_GB.html">VLC</a>, <a href="https://play.google.com/store/apps/details?id=net.sourceforge.servestream">ServeStream</a>,  ... and most Icecast clients.
- Supports sending of meta data (sending artist and title to the streaming client).
- Playlists are simple JSON files and data files are normal media (e.g. `.mp3`, `.nsv`) files on disk.
- Can be used as a stand-alone server or embedded in other Go projects.
- Supports HTTP basic user authentication.

Getting Started (standalone application)
----------------------------------------
You can download a pre-compiled package for Windows (win64) or Linux (amd64) [here](https://void.devt.de/pub/dudeldu).

You can also pull the latest docker image of DudelDu from [Dockerhub](https://hub.docker.com/r/krotik/dudeldu):
```
docker pull krotik/dudeldu
```

Create an empty directory, change into it and run the following to start DudelDu:
```
docker run --rm --user $(id -u):$(id -g) -v $PWD:/data -p 9091:9091 krotik/dudeldu -host 0.0.0.0 <playlist>
```
The container will have access to the current local directory and all subfolders.

### Demo

DudelDu comes with a demo playlist. After extracting DudelDu switch to the directory `examples/demo`. Run ./run_demo.sh (Linux) or run_demo.bat (Windows) to start the server.

Open a browser and view the `demo.html` in the `examples/demo` directory. To access the demo streams you are prompted for a username and password. The credentials are:
```
username: web
password: web
```
You can also point your favourite audio streaming client (e.g. VLC) to the streaming URL:
```
http://localhost:9091/bach/cello_suite1
```
The demo includes also a small video in the [Nullsoft Streaming Video](https://en.wikipedia.org/wiki/Nullsoft_Streaming_Video) format (NSV). To see it point a video streaming client (e.g. VLC) to:
```
http://localhost:9091/trailer/big_buck_bunny
```
Note: By default you can only reach the streams via localhost. Use the -host parameter with a host name or IP address to expose it to external network peers.

### Command line options
The main DudelDu executable has the following command line options:
```
DudelDu x.x.x
Usage of ./dudeldu [options] <playlist>
  -?	Show this help message
  -auth string
    	Authentication as <user>:<pass>
  -debug
    	Enable extra debugging output
  -fqs int
    	Frame queue size (default 10000)
  -host string
    	Server hostname to listen on (default "127.0.0.1")
  -loop
    	Loop playlists
  -port string
    	Server port to listen on (default "9091")
  -pp string
    	Prefix all paths with a string
  -shuffle
    	Shuffle playlists
  -tps int
    	Thread pool size (default 10)

Authentication can also be defined via the environment variable: DUDELDU_AUTH="<user>:<pass>"
```

Building DudelDu
----------------
To build DudelDu from source you need to have Go installed (go >= 1.12):

Create a directory, change into it and run:
```
git clone https://devt.de/krotik/dudeldu/ .
```

You can build DudelDu's executable with:
```
go build ./server/dudeldu.go
```

Building DudelDu as Docker image
--------------------------------
DudelDu can be build as a secure and compact Docker image.

- Create a directory, change into it and run:
```
git clone https://devt.de/krotik/dudeldu/ .
```

- You can now build the Docker image with:
```
docker build --tag krotik/dudeldu .
```

License
-------
DudelDu source code is available under the [MIT License](/LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.3.0](https://devt.de///compare/v1.2.0...v1.3.0) (2019-09-22)


### Features

* Adding instance specific debug logging ([77d225c](https://devt.de///commit/77d225c))



## [1.2.0](https://devt.de///compare/v1.1.0...v1.2.0) (2019-09-13)


### Features

* Playlists can now contain URLs ([523ac35](https://devt.de///commit/523ac35))



## 1.1.0 (2019-09-08)


### Features

* Initial commit ([33620bd](https://devt.de///commit/33620bd))


server/dudeldu.go
/*
 * DudelDu
 *
 * Copyright 2016 Matthias Ladkau. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the MIT
 * License, If a copy of the MIT License was not distributed with this
 * file, You can obtain one at https://opensource.org/licenses/MIT.
 */

/*
DudelDu main entry point for the standalone server.

Features:

- Supports various streaming clients: VLC, ServeStream, ... and most Icecast clients.

- Supports sending of meta data (sending artist and title to the streaming client).

- Playlists are simple json files and data files are normal media (e.g. .mp3) files on disk.

- Supports basic authentication.
*/
package main

import (
	"flag"
	"fmt"
	"log"
	"os"

	"devt.de/krotik/dudeldu"
	"devt.de/krotik/dudeldu/playlist"
)

// Global variables
// ================

/*
ConfigFile is the config file which will be used to configure DudelDu
*/
var ConfigFile = "dudeldu.config.json"

/*
Known configuration options for DudelDu
*/
const (
	ThreadPoolSize = "ThreadPoolSize"
	FrameQueueSize = "FrameQueueSize"
	ServerPort     = "ServerPort"
	ServerHost     = "ServerHost"
	PathPrefix     = "PathPrefix"
)

/*
DefaultConfig is the defaut configuration
*/
var DefaultConfig = map[string]interface{}{
	ThreadPoolSize: 10,
	FrameQueueSize: 10000,
	ServerPort:     "9091",
	ServerHost:     "127.0.0.1",
	PathPrefix:     "",
}

type consolelogger func(v ...interface{})

/*
Fatal/print logger methods. Using a custom type so we can test calls with unit
tests.
*/
var fatal = consolelogger(log.Fatal)
var print = consolelogger(func(a ...interface{}) {
	fmt.Fprint(os.Stderr, a...)
	fmt.Fprint(os.Stderr, "\n")
})

var lookupEnv func(string) (string, bool) = os.LookupEnv

/*
DudelDu server instance (used by unit tests)
*/
var dds *dudeldu.Server

/*
Main entry point for DudelDu.
*/
func main() {
	var err error
	var plf dudeldu.PlaylistFactory

	print(fmt.Sprintf("DudelDu %v", dudeldu.ProductVersion))

	auth := flag.String("auth", "", "Authentication as <user>:<pass>")
	serverHost := flag.String("host", DefaultConfig[ServerHost].(string), "Server hostname to listen on")
	serverPort := flag.String("port", DefaultConfig[ServerPort].(string), "Server port to listen on")
	threadPoolSize := flag.Int("tps", DefaultConfig[ThreadPoolSize].(int), "Thread pool size")
	frameQueueSize := flag.Int("fqs", DefaultConfig[FrameQueueSize].(int), "Frame queue size")
	pathPrefix := flag.String("pp", DefaultConfig[PathPrefix].(string), "Prefix all paths with a string")
	enableDebug := flag.Bool("debug", false, "Enable extra debugging output")
	loopPlaylist := flag.Bool("loop", false, "Loop playlists")
	shufflePlaylist := flag.Bool("shuffle", false, "Shuffle playlists")
	showHelp := flag.Bool("?", false, "Show this help message")

	flag.Usage = func() {
		print(fmt.Sprintf("Usage of %s [options] <playlist>", os.Args[0]))
		flag.PrintDefaults()
		print()
		print(fmt.Sprint("Authentication can also be defined via the environment variable: DUDELDU_AUTH=\"<user>:<pass>\""))
	}

	flag.Parse()

	if len(flag.Args()) != 1 || *showHelp {
		flag.Usage()
		return
	}

	// Check for auth environment variable

	if envAuth, ok := lookupEnv("DUDELDU_AUTH"); ok && *auth == "" {
		*auth = envAuth
	}

	laddr := fmt.Sprintf("%v:%v", *serverHost, *serverPort)

	print(fmt.Sprintf("Serving playlist %v on %v", flag.Arg(0), laddr))
	print(fmt.Sprintf("Thread pool size: %v", *threadPoolSize))
	print(fmt.Sprintf("Frame queue size: %v", *frameQueueSize))
	print(fmt.Sprintf("Loop playlist: %v", *loopPlaylist))
	print(fmt.Sprintf("Shuffle playlist: %v", *shufflePlaylist))
	print(fmt.Sprintf("Path prefix: %v", *pathPrefix))
	if *auth != "" {
		print(fmt.Sprintf("Required authentication: %v", *auth))
	}

	// Create server and listen

	plf, err = playlist.NewFilePlaylistFactory(flag.Arg(0), *pathPrefix)

	if err == nil {

		rh := dudeldu.NewDefaultRequestHandler(plf, *loopPlaylist, *shufflePlaylist, *auth)
		dds = dudeldu.NewServer(rh.HandleRequest)
		dds.DebugOutput = *enableDebug

		rh.SetDebugLogger(dds)

		defer print("Shutting down")

		err = dds.Run(laddr, nil)
	}

	if err != nil {
		fatal(err)
	}
}
